### PR TITLE
 zero_zilch_nada: remove ambiguous reference to empty

### PR DESCRIPTION
**Pull request by sudomateo**
In Go, there can be a difference between a type's zero value and its
empty value. For example, consider the following code snippet:

```go
func main() {
	var s1 []string
	fmt.Printf("%#v\n", s1)
	if s1 == nil {
		fmt.Println("s1 is equal to nil")
	}

	s2 := []string{}
	fmt.Printf("%#v\n", s2)
	if s2 == nil {
		fmt.Println("s2 is equal to nil")
	}
}
```

The variable `s1` will be set to its zero value (`nil`) whereas the
variable `s2` will be set to a slice of strings with zero elements. This
behavior makes the term "empty" ambiguous and may confuse students as they
continue learning about Go.

In addition to code updates, the exercise text was updated to introduce
the `var` keyword as well as various verbiage updates.

Updated the documentation for the nil and zero-values concepts to better
reflect the new `zero_zilch_nada` exercise content. Long term, I'd
probably recommend merging the two concepts into one but for now keeping
their content the same ensures that students will learn the relevant
content regardless of which concept they click on.

### CODE REVIEW

File: concepts/nil/about.md
Comment (ErikSchierboom) [inline concepts/nil/about.md:39]: I would probably leaving this out, as this document is about `nil` and not about zero values. If you'd like to link between concepts, I'd suggest using the Concept Widget as described in [this document](https://exercism.org/docs/building/markdown/widgets).

File: concepts/nil/about.md
Comment (ErikSchierboom) [inline concepts/nil/about.md:7]: Maybe consider adding backticks around the complex types and using their actual type name?

File: concepts/nil/about.md
Comment (iHiD) [inline concepts/nil/about.md:3]:
```suggestion
Go does not have a concept of empty/null/undefined for variable values. Variables declared without an explicit initial value default to the [zero value][zero_values] for their respective type.
```

I'd like to keep this bit in - it's useful for people from coming from other languages to understand there is no `null` (as `null` and `undefined` are quite different things)

File: concepts/nil/introduction.md
Comment (ErikSchierboom) [inline concepts/nil/introduction.md:1]: It looks like this file is just a copy of the `about.md` document, which is not exactly how it should be (see [the docs](https://exercism.org/docs/building/tracks/concept-exercises)). In short: the introduction should be a minimal explanation of the concept.

File: concepts/nil/introduction.md
Comment (ErikSchierboom) [inline concepts/nil/introduction.md:65]: If I look at the exemplar implementation, it looks like students don't need to know about this to succesfully finish the exercise. If true, could you omit this example? You can read more on this here: https://exercism.org/docs/building/tracks/concept-exercises

File: concepts/nil/introduction.md
Comment (ErikSchierboom) [inline concepts/nil/introduction.md:67]: Introduction documents should generally not link to external resources. Could you remove the link?

File: exercises/concept/zero-zilch-nada/.docs/introduction.md
Comment (ErikSchierboom) [inline exercises/concept/zero-zilch-nada/.docs/introduction.md:65]: If I look at the exemplar implementation, it looks like students don't need to know about this to succesfully finish the exercise. If true, could you omit this example? You can read more on this here: https://exercism.org/docs/building/tracks/concept-exercises

File: exercises/concept/zero-zilch-nada/.docs/instructions.md
Comment (ErikSchierboom) [inline exercises/concept/zero-zilch-nada/.docs/instructions.md:5]: I wonder if it would be possible to have this exercise depend on those other exercises. For example, one could teach booleans first in another exercise without that exercise having to depend on the zero concept.

File: concepts/zero-values/introduction.md
Comment (ErikSchierboom) [inline concepts/zero-values/introduction.md:65]: If I look at the exemplar implementation, it looks like students don't need to know about this to succesfully finish the exercise. If true, could you omit this example? You can read more on this here: https://exercism.org/docs/building/tracks/concept-exercises

File: concepts/zero-values/.meta/config.json
Comment (ErikSchierboom) [inline concepts/zero-values/.meta/config.json:3]: Maybe add yourself as the author?

File: concepts/zero-values/introduction.md
Comment (ErikSchierboom) [inline concepts/zero-values/introduction.md:67]: Introduction documents should generally not link to external resources. Could you remove the link?

File: concepts/zero-values/about.md
Comment (ErikSchierboom) [inline concepts/zero-values/about.md:7]: I would argue that this document does not need to describe `nil` values (as that is a different concept). If you'd like to link between concepts, I'd suggest using the Concept Widget as described in this document.

**Review by ErikSchierboom — APPROVED** [review]
Thanks a lot for working on this! I've left some comments to help explain what is expected in the various documents.

I don't know if you're interested, but this exercise could use a little love. In its current state it is, well, rather matter-of-factly. You just return a single value from the functions and there is no real story/theme around the functionality to give the students something fun to work with. Would you be interested in trying to add a bit more flavor to this exercise and make it more fun for students? If so, let me know, and I can help you with that.

File: exercises/concept/zero-zilch-nada/.docs/introduction.md
Comment (sudomateo) [reply exercises/concept/zero-zilch-nada/.docs/introduction.md]: Omitted the example!

File: concepts/zero-values/.meta/config.json
Comment (sudomateo) [reply concepts/zero-values/.meta/config.json]: Done!

File: concepts/zero-values/introduction.md
Comment (sudomateo) [reply concepts/zero-values/introduction.md]: Omitted the example!

File: concepts/zero-values/about.md
Comment (sudomateo) [reply concepts/zero-values/about.md]: Removed the `nil` values and added a Concept Widget.

File: concepts/zero-values/introduction.md
Comment (sudomateo) [reply concepts/zero-values/introduction.md]: Done!

**Comment by sudomateo** [comment]
Thank you for the feedback! I reworked the `zero-values` concept per your feedback. I also updated the `zero-zilch-nada` exercise to be a bit more involved for the student.

I'll do the same for the `nil` concept once I get more time to work on this. For now I'm leaving conversations open around the `nil` concept.

**Comment by ErikSchierboom** [comment]
> I also updated the zero-zilch-nada exercise to be a bit more involved for the student.

Maybe I'm missing something, but the instructions of the exercise look to be relatively unchanged. With more involved, I meant giving the exercise a little bit of story/theme, to have the tasks make sense for the student (and not just a list of fill-in methods). Note that this story/theme doesn't have to be very involved, it can be relatively simple. For some inspiration, I'd look at existing exercises like https://github.com/exercism/elixir/blob/main/exercises/concept/dna-encoding/.docs/instructions.md or https://github.com/exercism/elixir/blob/main/exercises/concept/high-school-sweetheart/.docs/instructions.md

**Comment by junedev** [comment]
@sudomateo As a story I can recommend the amusement park story: https://github.com/exercism/javascript/blob/main/exercises/concept/amusement-park/.docs/instructions.md

Write some introduction sentences about working for the amusement park IT and being in charge of data quality of the visitor data etc.

Then you can set up the tasks the following way:

- First Task: let the student write a NewVisitor function that creates and returns a visitor struct given some data as parameters. Describe in the task descriptions which fields the struct should have, e.g. name as string, age as int, address as map (street, city), etc. This task does not practice anything yet really, it is more for the setup.
- Then the following tasks can be things like "Determine whether the visitor provided a name" where you pass a visitor struct as parameter and the function e.g. "nameIsSet" returns true or false depending on the value of that field.
- I would see the above as the minimum to make this a bit more story-like. If you want to go further you can also mix it up with something like this task where you provide some map and the student needs to return some text: https://github.com/exercism/javascript/blob/main/exercises/concept/amusement-park/.docs/instructions.md#4-improve-the-ticket-status-response

I don't think it is necessary that the exercise covers all the types. Concentrate on a few tasks to cover the most common types. That's enough for the student to understand the idea.

If you go ahead with these types of tasks, don't forget to add structs as a prerequisite.

Erik says it is ok to change the slug and folder name to "amusement-park".

**Comment by ErikSchierboom** [comment]
> I don't think it is necessary that the exercise covers all the types. Concentrate on a few tasks to cover the most common types. That's enough for the student to understand the idea.

100% this ☝️

**Comment by sudomateo** [comment]
Thank you for the feedback @ErikSchierboom and @junedev. I've been a bit busy in and out of work the past week. I'll try to take a look at this in the next few days. I like the idea of making the exercise a bit more story-like. When I first opened this pull request, I was under the impression I should leave the exercise alone. It's nice to see that I can help shape the exercise to be more engaging.

**Comment by jmrunkle** [comment]
FYI: CI should be fixed at head if you do not mind syncing your fork with main and then rebasing this branch.

**Comment by sudomateo** [comment]
I'm going modify the `zero-zilch-nada` exercise and rename it to `census`. This `census` exercise will ask the user to create new residents with their necessary information, check whether or not required information is missing from a given resident, allow the counting of valid residents with all required information present, and allow residents to "move out" by setting their fields to their zero value. It's similar to the amusement park exercise with a bit more focus on zero values. I'll update this pull request once I finish the exercise. Apologies for the delay, last minute travel plans.

**Comment by ErikSchierboom** [comment]
> I'm going modify the zero-zilch-nada exercise and rename it to census. This census exercise will ask the user to create new residents with their necessary information, check whether or not required information is missing from a given resident, allow the counting of valid residents with all required information present, and allow residents to "move out" by setting their fields to their zero value. It's similar to the amusement park exercise with a bit more focus on zero values. I'll update this pull request once I finish the exercise. Apologies for the delay, last minute travel plans.

That sounds excellent.

File: concepts/nil/about.md
Comment (sudomateo) [reply concepts/nil/about.md]: Done!

File: concepts/nil/introduction.md
Comment (sudomateo) [reply concepts/nil/introduction.md:1]: Updated!

File: concepts/nil/introduction.md
Comment (sudomateo) [reply concepts/nil/introduction.md]: Done!

File: config.json
Comment (sudomateo) [inline config.json:230]: I only changed the `name` and `slug` for now. Perhaps I should also change the `prerequisites` as well?

File: exercises/concept/census/.docs/instructions.md
Comment (ErikSchierboom) [inline exercises/concept/census/.docs/instructions.md:1]: I really, really like this story and their tasks! Well done 👍

**Review by ErikSchierboom — COMMENTED** [review]
This looks great! I'll let one of the Go maintainers review the Go side of things.

Thanks a lot!

**Comment by ErikSchierboom** [comment]
There is a merge conflict.

File: exercises/concept/census/.docs/hints.md
Comment (sudomateo) [reply exercises/concept/census/.docs/hints.md]: Switched to widgets!

File: exercises/concept/census/.docs/instructions.md
Comment (sudomateo) [reply exercises/concept/census/.docs/instructions.md]: Used the word `process` instead.

File: exercises/concept/census/.docs/instructions.md
Comment (sudomateo) [reply exercises/concept/census/.docs/instructions.md:1]: Thank you kindly!

File: concepts/nil/about.md
Comment (sudomateo) [reply concepts/nil/about.md:1]: You're right. Both `nil` and `zero-values` are quite identical. Currently, they both point to the same exercise. Perhaps for now we leave the content as is and defer the work of differentiating their content until someone can update both the `nil` content and create a specific exercise for it? What do you think?

File: config.json
Comment (ErikSchierboom) [inline config.json]: Why did you add this entry? Was it to fix CI?

File: exercises/concept/census/.docs/hints.md
Comment (ErikSchierboom) [inline exercises/concept/census/.docs/hints.md:7]: For now, concept widgets render inline:

```suggestion
- Revisit the [concept:go/structs]() concept for information on how to create structs.
```

File: exercises/concept/census/.docs/hints.md
Comment (ErikSchierboom) [inline exercises/concept/census/.docs/hints.md:15]:
```suggestion
- Revisit the [concept:go/maps]() concept for information on how to determine if a map key exists.
```

File: exercises/concept/census/.docs/hints.md
Comment (ErikSchierboom) [inline exercises/concept/census/.docs/hints.md:22]:
```suggestion
- Revisit the [concept:go/zero-values]() concept for the zero value for each Go type.
```

File: exercises/concept/census/.docs/hints.md
Comment (ErikSchierboom) [inline exercises/concept/census/.docs/hints.md:28]:
```suggestion
- Revisit the [concept:go/range-iteration]() concept for information on how to range over a slice.
```

File: config.json
Comment (sudomateo) [reply config.json]: I did not add this. I merged `main` into this branch. This entry is already on `main` [here](https://github.com/exercism/go/blob/main/config.json#L237-L244).

File: concepts/nil/about.md
Comment (junedev) [reply concepts/nil/about.md:1]: I agree with @sudomateo this is hard to separate out in this context. What about having this exercise just teach one concept `zero-values` and have all the content we really need in there?

Then we don't have the duplication and we can still decide later on to add separate concept here or in the new "pointers" exercise (https://github.com/exercism/go/issues/1643).

Comment (ErikSchierboom) [reply concepts/nil/about.md:1]: Sounds good.

File: exercises/concept/census/.docs/hints.md
Comment (junedev) [inline exercises/concept/census/.docs/hints.md:5]: @ErikSchierboom Does it make sense to use widgets inside of a line of text? I thought they render as rather big boxes once they are actually implemented. I would have expected this to be a normal relative link to e.g. `/tracks/go/concepts/structs`.

File: exercises/concept/census/.docs/instructions.md
Comment (junedev) [inline exercises/concept/census/.docs/instructions.md:37]: `make` should be only called with 1 argument in case of map

File: exercises/concept/census/.docs/instructions.md
Comment (junedev) [inline exercises/concept/census/.docs/instructions.md:22]: no comma at end if the bracket is not on a new line

File: exercises/concept/census/census_test.go
Comment (junedev) [inline exercises/concept/census/census_test.go:143]: A separate test function for each function/method would give much better test output for the student. Not sure the rules of the Go track also allow the current version. I asked on Slack.

**Comment by junedev** [comment]
> Nice! @junedev could you look at the Go side of things?

Done

File: exercises/concept/census/census_test.go
Comment (ekingery) [reply exercises/concept/census/census_test.go:143]: Thanks for pointing this out, @junedev. As written, the entire test will be returned for every failure. In order to parse out the individual subtests properly, this test would need to be refactored to loop through the test cases immediately after they are defined. Here is the spec: https://github.com/exercism/go-test-runner#subtests

File: exercises/concept/census/.docs/instructions.md
Comment (sudomateo) [reply exercises/concept/census/.docs/instructions.md]: In this case, yep!

File: exercises/concept/census/.docs/hints.md
Comment (sudomateo) [reply exercises/concept/census/.docs/hints.md:5]: I'm happy to implement either way here.

File: exercises/concept/census/census_test.go
Comment (sudomateo) [reply exercises/concept/census/census_test.go:206]: Updated to use multiple test functions.

File: concepts/nil/about.md
Comment (sudomateo) [reply concepts/nil/about.md:1]: @junedev Does this mean we should remove the `nil` concept entirely and move its unique information into the `zero-value` concept?

Comment (junedev) [reply concepts/nil/about.md:1]: Exactly!

File: concepts/zero-values/about.md
Comment (junedev) [reply concepts/zero-values/about.md:21]: Sorry, I got confused with the nil concept. Can you double check some short info on structs is also available in the introduction of the concept and exercise. I didn't see it there.

File: concepts/zero-values/about.md
Comment (junedev) [inline concepts/zero-values/about.md:32]: You can remove the main function from those examples if you want. But in any case this is missing a line below that shows the actual result, it is usually written like this on the Go track:  
`// Output: Zero value ....`

Sorry for all the nit picks, in general you did great work and I really appreciate your help!

File: concepts/nil/about.md
Comment (sudomateo) [reply concepts/nil/about.md]: Removed the `nil` concept and updated the `zero-value` concept with its information.

File: concepts/zero-values/about.md
Comment (sudomateo) [reply concepts/zero-values/about.md:21]: Done!

File: exercises/concept/census/.docs/hints.md
Comment (sudomateo) [reply exercises/concept/census/.docs/hints.md]: I used normal relative links to the relevant concept.

File: concepts/zero-values/about.md
Comment (sudomateo) [reply concepts/zero-values/about.md:33]: Never a nit pick! Thank you for the feedback.

**Review by junedev — APPROVED** [review]

**Comment by junedev** [comment]
@sudomateo The past thing that remains is getting the CI to pass. 

For the prerequisites you need to write "string-formatting" instead of "strings" for now. (It will be changed later when "strings" becomes available. Additionally you need to remove "pointers" for now as it does not exists yet. You can leave a comment here https://github.com/exercism/go/issues/1643 that it needs to be added again later.

**Comment by junedev** [comment]
As for the other CI failures, can you try to use "package census" instead of "package census_test" for the test file? This is how it was done in the other concept exercises. (Not sure this will fix the issue though.)

**Review by ekingery — APPROVED** [review]
Rubber stamping based on the other reviews / fixes and the tests passing.

**Comment by sudomateo** [comment]
Fixed the merge conflict.

**Comment by ekingery** [comment]
> Fixed the merge conflict.

By the way, thanks for your work on this exercise, @sudomateo! Looks great.

**Comment by sudomateo** [comment]
> > Fixed the merge conflict.
> 
> By the way, thanks for your work on this exercise, @sudomateo! Looks great.

Anytime! I'm happy to help!

**Comment by ErikSchierboom** [comment]
Yeah, great work @sudomateo!

**Pull request merged**
